sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -613,13 +633,62 @@ def _blocking_request(self, request):
     if response.error:
       raise RuntimeError(response.error)
     else:
+      # TODO mxm here we could update the cache token
       return response
 
   def _next_id(self):
     self._last_id += 1
     return str(self._last_id)
 
 
+class CachingGrpcStateHandler(GrpcStateHandler):
+  """ TODO mxm """
+
+  def __init__(self, state_stub, global_state_cache):
+    super(CachingGrpcStateHandler, self).__init__(state_stub)
+    self._state_cache = global_state_cache
+
+  def blocking_get(self, state_key, coder):
+    cache_tokens = self._context.cache_tokens
+    if not cache_tokens:
+      # no cache tokens, can't do a lookup in the cache
+      return super(CachingGrpcStateHandler, self).blocking_get(
+          state_key, coder)
+    cache_state_key = CachingGrpcStateHandler.convert_to_cache_key(state_key)
+    value = self._state_cache.get(cache_state_key, cache_tokens)
+    if not value:
+      # Cache miss, need to retrieve from the Runner
+      value = super(CachingGrpcStateHandler, self).blocking_get(
+          state_key, coder)
+      # TODO mxm uses always the first cache token for now
+      self._state_cache.put(
+          CachingGrpcStateHandler.convert_to_cache_key(state_key),
+          cache_tokens[0], value)
```

Comment:
I am not sure why we are always taking the first token: `cache_tokens[0]`

Reply:
The current protocol also return the cache token, are we not caching that token here?

Reply:
Implementation of this method suggests that you are passing only all tokens.

Reply:
As per last discussion on the mailing list, there will only be one active cache token for all user state. This token is provided by the Runner in the ProcessBundleRequest. We also decided to ignore the cache token returned by a get/put/clear request because we want to implement async communication and assume validity of the provided token for the duration of a bundle.

That said, if